pkg/proxy/dns/store.go

```diff
@@ -0,0 +1,241 @@
+package dns
+
+import (
+	"fmt"
+	"hash/fnv"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/golang/glog"
+
+	skymsg "github.com/skynetservices/skydns/msg"
+)
+
+type dnsStore struct {
+	cache.BulkIndexer
+	endpoints  client.EndpointsNamespacer
+	headlessMu sync.Mutex
+	isHeadless map[string]bool
+}
+
+type dnsEntry struct {
+	Namespace string
+	Name      string
+	skymsg.Service
+}
+
+var _ cache.Store = &dnsStore{}
+
+func makeSkyDNSService(host string, port int, key string, group string) skymsg.Service {
+	return skymsg.Service{
+		Host: host,
+		Port: port,
+
+		Priority: 10,
+		Weight:   10,
+		Ttl:      30,
+
+		Text: "",
+		Key:  key,
+
+		Group: group,
+	}
+}
+
+func makeEntriesForService(service *api.Service) []interface{} {
+	entries := make([]interface{}, len(service.Spec.Ports)+1)
+	srvPath := fmt.Sprintf(srvPathFormat, service.Name, service.Namespace)
```

Comment:
The reason why I don't just store services directly:

1. If you want proper reverse DNS (which seems to be desired), you need to be able index the endpoint information as well.  However, each `Endpoints` object has multiple IP addresses associated with it, so we can't just store endpoints objects -- we have to store individual records somehow.

2. It makes certain lookup operations easier -- for instance, instead of having to retrieve service information, then loop over ports comparing for the correct name, we can just add an index on the SkyDNS `Key` field, and jump right to the correct entries.  Instead of having to get the service from the cache, check the service to see if it's headless, and then query for endpoints, we just get all entries associated with a namespace/name pair.

3. All necessary information is cached (as opposed to the OpenShift resolver, which doesn't cache endpoint information)

4. It seems to be faster (lookup-wise) than transformation on the fly, in some basic informal tests